src/bucket.ts

```diff
@@ -2039,15 +2040,32 @@ class Bucket extends ServiceObject {
       });
     };
 
-    this.getFiles(query)
-      .then(([files]) => {
+    (async () => {
+      try {
+        let promises = [];
         const limit = pLimit(MAX_PARALLEL_LIMIT);
-        const promises = files!.map(file => {
-          return limit(() => deleteFile(file));
-        });
-        return Promise.all(promises);
-      })
-      .then(() => callback!(errors.length > 0 ? errors : null), callback!);
+        const filesStream = this.getFilesStream(query);
+
+        for await (const curFile of filesStream) {
+          if (promises.length >= MAX_QUEUE_SIZE) {
+            await Promise.all(promises);
```

Comment:
Once we hit the queue limit, all execution gets stopped until all the pending promises have finished. This technically works, but doesn't maximize the benefit of streams and results in lower performance.

When I was trying to find a workaround for this issue, I wrote this example transform stream for limiting execution of a stream: https://gist.github.com/timscottbell/364e2683354683c05cdfd93c708e9072. This is not production-ready code or fully tested, but more to give an idea of the direction I thought this solution would go in.

Our current workaround is this:
```
  async deleteBucket(name: string) {
    const bucket = StorageManagerGoogle._bucket(name, config.CloudProjectId);

    const bucketExists = await StorageManagerGoogle._exists(bucket);

    if (!bucketExists) {
      logger.warn(`Bucket [${name}] does not exist, cannot delete it`);

      return;
    }

    while (true) {
      const [files] = await bucket.getFiles({
        autoPaginate: false,
        maxResults: 1000,
      });

      if (files.length === 0) {
        break;
      }

      await Promise.all(
        files.map(async (file: File) => {
          try {
            await file.delete({ignoreNotFound: true});
          } catch (error) {
            const {response} = error;

            if (
              !response ||
              response.statusCode !== 409 ||
              response.statusMessage !== 'Conflict'
            ) {
              throw error;
            }
          }
        })
      );
    }

    await bucket.delete();
  }
  ```
  
  which looks very surprisingly similar to the approach here. (We opted for the less performant solution over stream solution because some team members have less familiarity with streams, and it was a workaround we only wanted to commit temporarily).

Reply:
I've added some other improvements to our backlog around `deleteFiles` performance in addition to some other stream related work we have planned for the next major version. 

You are correct, we essentially back pressure the stream once the queue is full. I did this intentionally to not hold thousands / tens of thousands of objects in the queue at the same time.  Hopefully this is enough of a reduction in memory for the time being. I don't think this will degrade performance as even the previous version only allowed 10 concurrent promises executing even if the queue contained 100,000 items.

If this doesn't help to resolve the issue you were seeing please feel free to update the open ticket and we will of course be happy to take a look.